File: artifacts/api-server/src/routes/jobs.ts
import { Router, type IRouter } from "express";
import { eq, and, ilike, or } from "drizzle-orm";
import { db, jobsTable } from "@workspace/db";
import {
  ListJobsQueryParams,
  ListJobsResponse,
  GetJobResponse,
  GetTrendingJobsResponse,
} from "@workspace/api-zod";
import { stripNulls, stripNullsArray } from "../lib/nullstrip";

const router: IRouter = Router();

router.get("/jobs/trending", async (_req, res): Promise<void> => {
  const jobs = await db
    .select()
    .from(jobsTable)
    .where(eq(jobsTable.isTrending, true))
    .limit(10);

  res.json(GetTrendingJobsResponse.parse(stripNullsArray(jobs)));
});

router.get("/jobs", async (req, res): Promise<void> => {
  const params = ListJobsQueryParams.safeParse(req.query);
  if (!params.success) {
    res.status(400).json({ error: "invalid_params", message: params.error.message });
    return;
  }

  const conditions = [];
  if (params.data.sector) {
    conditions.push(eq(jobsTable.sector, params.data.sector));
  }
  if (params.data.category) {
    conditions.push(ilike(jobsTable.category, `%${params.data.category}%`));
  }
  if (params.data.state) {
    conditions.push(eq(jobsTable.state, params.data.state));
  }
  if (params.data.educationLevel) {
    conditions.push(eq(jobsTable.educationLevel, params.data.educationLevel));
  }
  if (params.data.search) {
    conditions.push(
      or(
        ilike(jobsTable.title, `%${params.data.search}%`),
        ilike(jobsTable.organization, `%${params.data.search}%`)
      )
    );
  }

  const jobs = await db
    .select()
    .from(jobsTable)
    .where(conditions.length > 0 ? and(...conditions) : undefined)
    .orderBy(jobsTable.createdAt);

  res.json(ListJobsResponse.parse(stripNullsArray(jobs)));
});

router.get("/jobs/:id", async (req, res): Promise<void> => {
  const raw = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const id = parseInt(raw, 10);

  if (isNaN(id)) {
    res.status(400).json({ error: "invalid_params", message: "Invalid job ID" });
    return;
  }

  const [job] = await db
    .select()
    .from(jobsTable)
    .where(eq(jobsTable.id, id));

  if (!job) {
    res.status(404).json({ error: "not_found", message: "Job not found" });
    return;
  }

  res.json(GetJobResponse.parse(stripNulls(job)));
});

export default router;

File: artifacts/api-server/src/routes/jobs.rs
use rocket::serde::json::Json;
use rocket::{get, routes, Route, State};
use rocket::form::FromForm;
use sqlx::PgPool;
use crate::errors::{internal_error, not_found_error, ApiResult};
use crate::models::Job;

const JOB_SELECT: &str = r#"
    SELECT id, title, organization, sector, category, location, state,
        education_level, vacancies, salary_min, salary_max, salary_text,
        application_deadline, application_url, logo_url, is_trending, is_urgent,
        status, description, eligibility,
        COALESCE(tags, ARRAY[]::text[]) AS tags,
        posted_at, created_at
    FROM jobs
"#;

#[derive(FromForm)]
pub struct JobQuery {
    sector: Option<String>,
    category: Option<String>,
    state: Option<String>,
    #[field(name = "educationLevel")]
    education_level: Option<String>,
    search: Option<String>,
}

#[get("/jobs/trending")]
pub async fn trending_jobs(pool: &State<PgPool>) -> ApiResult<Vec<Job>> {
    let sql = format!(
        "{} WHERE is_trending = true ORDER BY created_at LIMIT 10",
        JOB_SELECT
    );
    let jobs = sqlx::query_as::<_, Job>(&sql)
        .fetch_all(pool.inner())
        .await
        .map_err(internal_error)?;
    Ok(Json(jobs))
}

#[get("/jobs?<q..>")]
pub async fn list_jobs(pool: &State<PgPool>, q: JobQuery) -> ApiResult<Vec<Job>> {
    let search_pattern = q.search.as_deref().map(|s| format!("%{}%", s));
    let sql = format!(
        r#"{} WHERE ($1::text IS NULL OR sector = $1)
           AND ($2::text IS NULL OR category ILIKE $2)
           AND ($3::text IS NULL OR state = $3)
           AND ($4::text IS NULL OR education_level = $4)
           AND ($5::text IS NULL OR (title ILIKE $5 OR organization ILIKE $5))
           ORDER BY created_at"#,
        JOB_SELECT
    );
    let jobs = sqlx::query_as::<_, Job>(&sql)
        .bind(q.sector)
        .bind(q.category.as_deref().map(|c| format!("%{}%", c)))
        .bind(q.state)
        .bind(q.education_level)
        .bind(search_pattern)
        .fetch_all(pool.inner())
        .await
        .map_err(internal_error)?;
    Ok(Json(jobs))
}

#[get("/jobs/<id>")]
pub async fn get_job(pool: &State<PgPool>, id: i32) -> ApiResult<Job> {
    let sql = format!("{} WHERE id = $1", JOB_SELECT);
    let job = sqlx::query_as::<_, Job>(&sql)
        .bind(id)
        .fetch_optional(pool.inner())
        .await
        .map_err(internal_error)?
        .ok_or_else(|| not_found_error("Job not found"))?;
    Ok(Json(job))
}

pub fn routes() -> Vec<Route> {
    routes![trending_jobs, list_jobs, get_job]
}
